### PR TITLE
WIP: Use UBOs in project module

### PR DESCRIPTION
**Pull request by felixpalmer**
#### Background

Status:

- Requires luma 9.0.12
- Most shaders & layers working
- `project64` module also ported
- All tests passing

#### Change List
- Update `project` module to use `projectUniforms` UBO
- Update references in all shaders, e.g. `project_uDevicePixelRatio` -> `project.devicePixelRatio`
- Remove `project` from `DEFAULT_MODULES` as it needs to be passed in the bindings
- `project64` now passes fp64 matrix as a `mat4` uniforms, rather than `vec2[16]`
- Test updates

### CODE REVIEW

File: modules/core/src/effects/lighting/camera-light.ts [hunk 1]
```diff
@@ -8,15 +8,15 @@ export default class CameraLight extends PointLight {
...
+    const {cameraPosition} = getUniformsFromViewport({
       viewport,
       modelMatrix,
       coordinateSystem,
       coordinateOrigin
     });
     projectedLight.color = this.color;
     projectedLight.intensity = this.intensity;
-    projectedLight.position = project_uCameraPosition;
+    projectedLight.position = cameraPosition;
```
Comment (ibgreen) [inline modules/core/src/effects/lighting/camera-light.ts:19]: Thought: I wonder if we should add a coordinate system suffix to position parameters. E.g. if this is the so called "common space":

```suggestion
    projectedLight.positionCS = cameraPositionCS;
```

File: modules/core/src/shaderlib/project/project.glsl.ts [hunk 2]
```diff
@@ -36,23 +36,28 @@ ${COORDINATE_SYSTEM_GLSL_CONSTANTS}
...
-uniform mat4 project_uModelMatrix;
-uniform mat4 project_uViewProjectionMatrix;
-uniform vec2 project_uViewportSize;
-uniform float project_uDevicePixelRatio;
-uniform float project_uFocalDistance;
-uniform vec3 project_uCameraPosition;
-uniform vec3 project_uCoordinateOrigin;
-uniform vec3 project_uCommonOrigin;
-uniform bool project_uPseudoMeters;
+uniform project32Uniforms {
```
Comment (ibgreen) [inline modules/core/src/shaderlib/project/project.glsl.ts:39]: Nit: Can we keep module and UBO name consistent? Do we have a good reason to diverge?
```suggestion
uniform projectUniforms {
```

File: modules/core/src/shaderlib/project/project.glsl.ts [hunk 3]
```diff
@@ -162,7 +168,7 @@ vec4 project_offset_(vec4 offset) {
 //
 vec2 project_mercator_(vec2 lnglat) {
   float x = lnglat.x;
-  if (project_uWrapLongitude) {
+  if (project.autoWrapLongitude) {
```
Comment (ibgreen) [inline modules/core/src/shaderlib/project/project.glsl.ts:171]: Why rename it?

**Review by ibgreen — APPROVED** [review]
Very nice, this is a big part of the lift required to make deck.gl WebGPU ready.

> Hacky implementation as luma doesn't handle module dependencies

I am not sure exactly what that means, but is it something we can address in luma.gl to keep this complicated part of deck.gl a little cleaner?

File: modules/core/src/shaderlib/project/project.glsl.ts (hunk 2 above)
Comment (felixpalmer) [reply modules/core/src/shaderlib/project/project.glsl.ts:39]: This is what I meant by "hacky"...

The `project32` module depends on `project`. I would like to use `uniform projectUniforms {...` but currently that won't work in luma as the `ShaderInputs` doesn't traverse the module depencies. So as I first step I tried to port the shaders with workarounds [like this](https://github.com/visgl/deck.gl/pull/8782/files#diff-9dc0b8596c862e17e7ef398d73fd9f6e0c1ea1865c0d8e26ead182080c2b3956R51-R53), but the view to then see what is required in luma to land this cleanly.

If you'd like to look into how to solve this from the luma side, that would be appreciated. We will have this issue with other modules also

**Comment by felixpalmer** [comment]
@Pessimistress @ibgreen landing this will be a breaking change for any custom shaders that reference `project_uXXXX` uniforms. Are we OK with this or do we want to consider automatically rewriting shaders by detecting `"project_u"` strings and replacing? We could warn that their shaders are using the old syntax and then force the upgrade in 9.1

**Comment by ibgreen** [comment]
> Are we OK with this or do we want to consider automatically rewriting shaders by detecting "project_u" strings and replacing? We could warn that their shaders are using the old syntax and then force the upgrade in 9.1

I hadn't even thought about the possibility of making the uniform buffer change backwards compatible, but it would of course be nice. I recommend looking into the existing functionality in ShaderModules, perhaps it works as is, or can be made to work with small changes:

image

File: modules/core/src/shaderlib/project/project.glsl.ts (hunk 2 above)
Comment (felixpalmer) [reply modules/core/src/shaderlib/project/project.glsl.ts:39]: Implemented in luma: https://github.com/visgl/luma.gl/pull/2073

File: modules/core/src/effects/lighting/camera-light.ts (hunk 1 above)
Comment (felixpalmer) [reply modules/core/src/effects/lighting/camera-light.ts:19]: Perhaps, but it is out of scope for this PR - there are enough changes here as-is

**Review by Pessimistress — APPROVED** [review]

**Comment by ibgreen** [comment]
Epic!

**Comment by donmccurdy** [comment]
Hi, just checking whether @felixpalmer's question in https://github.com/visgl/deck.gl/pull/8782#issuecomment-2051394539 was resolved — are we comfortable with this PR going into a deckgl v9.0.x patch release, or should we hold it for v9.1? I'm working on publishing v9.0.15 now. Thanks!

**Pull request merged**